### ISSUE TITLE
[Android] dnscrypt-proxy doesn't work on Android 10

### ISSUE THREAD

**Issue by iWARR** [DNSCrypt/dnscrypt-proxy#1149]
### ISSUE

dnscrypt-proxy on Android doesn't work on new Android 10 (Q).

dnscrypt-proxy can start, report about connection, but can't do the job.
The behaviour looks like it's appears restricted on system level or I don't know what...

### iNFO

Seems like some new changes in Andoid DNS daemon stop dnscrypt-proxy from working.

Let's have a look at the option:

Settings -> Network & Internet -> Private DNS

	( ) Off
	( ) Automatic
	(•) Private DNS provider hostname : localhost

Only the names can be accepted here (not the IPs). This option was designed for entering the names of alternative DNS servers that support DoT (DNS-over-TLS). There are almost no such servers over the Internet (let alone trusted ones). It is clear that when dnscrypt-proxy used, then this nonsense does not need anyone. But exactly this thing (IMO) prevents dnscrypt-proxy from working.

Values (•) Off, or (•) Automatically do not make sense here at all, because the servers of the "not evil" corporation will be used, or the mobile operator's defaults.

Since the OS uses a third-party DNS service (dnscrypt-proxy), in order to avoid interference and leaks, the obvious decision was made to specify there a value: `localhost`
The problem was quickly solved and forgotten before Android 10.

Now this trick doesn't work. From what I can assume that in Android 9 this option was experimental and optional, and DNS daemon in the Android 10 has become closely tied to this option and does not let run dnscrypt-proxy. I don't know the workaround for this yet.

Moreover, I know for sure that the dnscrypt-proxy binary is in the ready state. The dnscrypt-proxy connection logs say about readiness and finding the right encrypted servers according to my settings. But "Private DNS" option in Android 10 option does not allow dnscrypt-proxy to start working at system level as usual.

May be I'm wrong about a reason of the issue...
Any suggestions how this should be fixed?

**Comment by iWARR** [comment]
Just tested "InviZible Pro" for the first time. And it works...
I don't know how to explain this. May be the app uses more advanced IPTABLES rules or because it can use a Pseudo-VPN to bypass descibed issue...

**Comment by dapphp** [comment]
To continue to use dnscrypt-proxy on Android 10 you want to turn this option off. This way, the OS uses plain old DNS, which should be intercepted and redirected to your local copy of dnscrypt-proxy (depending on how it is installed).

dnscrypt-proxy is a proxy, so you don't want to be sending DoT queries directly to it which is what happens when you set Private DNS to "localhost".

**Comment by iWARR** [comment]
@dapphp 

I'm sure you have a reason. "InviZible Pro" works exactly as you said, using `(•) Off` setting.

But there was no luck with original dnscrypt-proxy pack in some case. 

For launching dnscrypt-proxy bundle I used to:
- init.d support
- Autorun scripts to start it when booting OS
- "Override DNS" app to override 127.0.0.1:53 globally + hard IPTABELS contol to prevent agressive IPTABLES resetting from the DNS daemon side of Android OS. "Override DNS" was always the only app that did this job effectively.

**Comment by iWARR** [comment]
Do you have dnscrypt-proxy working on Android 10?

**Comment by iWARR** [comment]
@**jedisct1**
If somebody will confirm dnscrypt-proxy works on Android 10, you may close this thread.
I don't know what's happen.
May be "Override DNS" is not ready for the Android 10 yet, or smth on my side...
All the time I see changes in (Q) there and there...

**Comment by dapphp** [comment]
Not on Android 10. I am still on LineageOS 16 (Android 9) but saw this post and know that to use dnscrypt-proxy the old way you would need to turn off "Private DNS".

I would check these things:

1. After booting Android, confirm that dnscrypt-proxy is running (make sure init script or auto-startup method is working)
2. Test DNS resolution. With Private DNS off, is Android using plain DNS to whatever resolver it got from DHCP or cellular?
3. Can you manually run iptables rules to redirect DNS to dnscrypt-proxy? Does Android 10 even use iptables anymore? Maybe nftables?

**Comment by iWARR** [comment]
According to my experience on Android 10:

1. Confirmed.

2. When (•) Off or (•) Automatic (doesn't matter):
 - Mobile Network: Mobile operator's DNS (Nework stuck)
- Wi-Fi: Captive portal, Time and nothing more (Nework stuck)

3. I think IPTABLES can't appear discontinued now by Google or it will be a Global Network apocalypse. "Override DNS" was always responsible for IPTABLES and overriding 127.0.0.1:53

**Comment by iWARR** [comment]
Also now I have 2 phones in my hands with Android 9 (Resurrection Remix OS, also LOS-based). As always, dnscrypt-proxy WORKS as usual.

BUT! If I set:

localhost = no DNS leak
(•) Off = DNS leak

That's why long time ago I decide to use `localhost` value for this option.

Checked 2 TIMES + REBOOT on Mobile Network, dnsleaktest.com

**Comment by dapphp** [comment]
> Also now I have 2 phones in my hands with Android 9 (Resurrection Remix OS, also LOS-based). As always, dnscrypt-proxy WORKS as usual.
> 
> BUT! If I set:
> 
> localhost = no DNS leak
> (•) Off = DNS leak
> 
> That's why long time ago I decide to use `localhost` value for this option.
> 
> Checked 2 TIMES + REBOOT on Mobile Network, dnsleaktest.com

Are these devices rooted?

If you turn that setting off and go to dnsleaktest.com, it should show your dnscrypt resolver IF dnscrypt-proxy was working as your main resolver.

What method is your init script using to redirect DNS traffic to dnscrypt-proxy?

**Comment by iWARR** [comment]
I'm always hated Goolag for things they do with DNS management in Android. And the new "Private DNS" feature was not an exception. 

> Are these devices rooted? 

Yes. All of these 3 are ROOTED (Magisk)

> If you turn that setting off and go to dnsleaktest.com, it should show your dnscrypt resolver IF dnscrypt-proxy was working as your main resolver.

As I said before, you have a sence. But in fact, it's not happen.

> What method is your init script using to redirect DNS traffic to dnscrypt-proxy?

Android 9 - Classic init.d method, working since 2015 until Android 9
Android 10 - Magisk module, a new pathes for the scripts (classic init.d method doesn't work any more)

**Comment by dapphp** [comment]
Seems like an Android/configuration issue.

From a terminal on the device, if `/path/to/dnscrypt-proxy -resolve github.com` works, then technically dnscrypt-proxy works on Android 10, but the OS needs to be configured properly. Turn off their "Private DNS" option and then figure out what it takes on Android 10 to hijack DNS traffic.

**Comment by iWARR** [comment]
I can't check this now, because I deleted stock bundle and install "InviZible Pro".
But I can say I saw in connection log that 2 my DNSCrypt servers were found, but I'm afraid the resolving didn't work (something shutted down dnscrypt-proxy from working on a system level and resolving was impossible). Otherwise my browser will start working immidiately.

I'm sorry I don't have the official bundle installed at present time... There was no any answer on my question for a long time (I waited for it all the day).

**Comment by iWARR** [comment]
But I'm still interesting in knowledge why all of these sad things happen. Thank you for you time, my friend. Appreciated.

**Comment by iWARR** [comment]
I don't wanna take much of your time, but if you agree to wait a 10-15 min., I can reinstall again and we will continue.

**Comment by iWARR** [comment]
@**jedisct1**
I'm always wanted ask you, is it possible to assign a separate UID to the process of dnscrypt-proxy for Android ? Because it's always stay inside [ 0 ] root process (Applications with ROOT access) that is not good for privacy and security. If so, we can disable Network access for the root process [ 0 ] inside AFWall+ and grant access only for dnscrypt UID [XXX].

I can post a screenshot if you want.

**Comment by dapphp** [comment]
Sure thing. Look at my Github profile for a website link that will have contact info for me. Feel free to email me directly. I can try to share my ways of getting dnscrypt-proxy to run on Android

Regarding your last question, yes, you can drop root access and have dnscrypt-proxy switch to a different user by adding `user_name = 'user'` to your dnscrypt-proxy.toml config file.

**Comment by iWARR** [comment]
> I can try to share my ways of getting dnscrypt-proxy to run on Android

I'm also nice to meet you :)

> to a different user 

I mean not user, but number of the UID system process.

**Comment by iWARR** [comment]
![DNSCrypt_UID](https://user-images.githubusercontent.com/33152262/72376554-09162880-3706-11ea-875b-fc29545f2d43.png)

**Comment by iWARR** [comment]
![DNSCrypt_UID](https://user-images.githubusercontent.com/33152262/72377848-29df7d80-3708-11ea-8351-d08e3131ab7d.png)

**Comment by jkle112** [comment]
InviZible Pro installed in android 10
1.need close Private DNS
2.start dnscrypt-proxy in InviZible Pro,if notice find a new version ,click update,it will auto upgrade and restart.
3.most devices are not rooted,need used VPN mode ,if used other VPN ,only changed to  proxy mode

**Comment by iWARR** [comment]
I inspected dnscrypt-proxy part of the "InviZible Pro" and can't recommend this app. A present versions, at least. There are very obsolete parameters and mistakes inside the toml config file and it's not corresponding to the modern dnscrypt-proxy settings. Some of settings in the "InviZible Pro" are not supported at all. Futhemore, if you will try to adopt modern **stock** config to this app (pathes and filenames) dnscrypt-proxy part will not work and crash.

**Comment by iWARR** [comment]
Some additional info for the specialists.

**DNS Resolver - Changes in Android 10**
https://source.android.com/devices/architecture/modular-system/dns-resolver

**Comment by iWARR** [comment]
What is bad in Android "traditionally" (all versions): Google DNS daemon is not under your control and doing in fact with malware methods, but **on system hardcoded level**.

Completely unstoppable (even with AFWall+ blocking they can bypass firewall restrictions):

- **www.google.com** - appears in DNS several times immidiatly after every Network connection
- **Time** (server can be changed)
- **Captive portal** (unstoppable even with ADB disabling (!), but server can be changed in ADB)
CP default host: `connectivitycheck.gstatic.com` - alternative & trusted can be used instead
A principles of the Captive portal functioning look like a MiTM.
Furthermore, it uses plain http (not https), Port:80 - it's just horrible
Even if you disable "Captive portal" system app OS component using MAT, DNS queries to `connectivitycheck.gstatic.com` will be still alive!
Captive portal start calling to Google immidiately after (and during) every Network connection (WiFi, Mobile) and operating like a separate DNS connectivity alongside already hardened (crypted) regular DNS. It's unacceptable even for "connectivity checking" proposes.

So, in fact, we need encrypted channels and dnscrypt-proxy support for the Captive portal connectivity (for not often connections in the public places, hotels, etc.). In that way we can try switching Android to the new trusted+encrypted channels for Captive connections in case we can't (by Goolag design) disable it comletely or Turn off it temporary when captive portal is not needed). 

I'm accentuate again, the Captive portal can make (and making) connections any time you start to use  Internet access.

How Captive Portals Interfere With Wireless Security and Privacy (2017)
https://www.eff.org/ru/deeplinks/2017/08/how-captive-portals-interfere-wireless-security-and-privacy

It's so sad is nobody concerned for years...
All examples above are relative to DNS, encription and secutity questions, so this is not "offtopic". (I beleive).

**Comment by iWARR** [comment]
I'm reinstall dnscrypt-proxy over again, but I'm still can't run it on Android 10.

@**dapphp**
I will have a contact with you as soon as I will have a time to be ready to our discussion.
(I wanna fix some issues by myself for a while).

**Comment by Gedsh** [comment]
> I inspected dnscrypt-proxy part of the "InviZible Pro" and can't recommend this app. A present versions, at least. There are very obsolete parameters and mistakes inside the toml config file and it's not corresponding to the modern dnscrypt-proxy settings.

dnscrypt-proxy has backward compatibility with old settings. This is not break functionality. All features works properly.

> Some of settings in the "InviZible Pro" are not supported at all.

As an InviZible developer, I use the options that are necessary for dnscrypt-proxy to work correctly. I implement the user interface only for options that require frequent changes, and not for everyone.

>  if you will try to adopt modern stock config to this app (pathes and filenames) dnscrypt-proxy part will not work and crash.

Do not touch the internal InviZible files, and everything will work correctly.

It takes time to adapt the InviZible configuration to the modern dnscrypt-proxy configuration. Be patient.

**Comment by iWARR** [comment]
@dapphp 
I sent you email yesterday, but still no answer.

**Comment by iWARR** [comment]
Seems I can guess where it stuck.

Google have provided a new restrictions in the Android 10:  the apps don't allowed to run in the background by default any more. So I decided to give a correspondent permission to binary using ADB command:

`cmd appops set /system/bin/dnscrypt-proxy RUN_IN_BACKGROUND allow`

Unfortunately with no luck, because the dnscrypt-proxy binary have none UID for user 0 :

`Error: No UID for /system/bin/dnscrypt-proxy in user 0`

Seems like this is another one reason to provide a separated UID for dnscrypt-proxy process : see my AFWall+ screenshot and notes above.

Anybody know how to make dnscrypt-proxy running in the background on Android 10?

**Comment by iWARR** [comment]
**Note:** _my dnscrypt-proxy is healthy itself and can resolve DNS through ADB commands._

**Comment by iWARR** [comment]
Updated to **v2.0.38** and now I can note significant inprovement with fallback resolvers and first start.
Both phones (Android 10) connected immidiately and fast.
All necessary logs created for seconds.

Private DNS: (•) Off (as it should be), no leaks

I've found all last changelog very useful.
Thanks for attention to the Windows service installation procedure, inline comments and all reliability improvements.
Exellent work! Exellent version!
